### DISCUSSION TITLE
Best practices for storing additional data

### DISCUSSION THREAD

**Discussion by snimavat** [apache/iotdb#12212] [Q&A]
What is general best practice to store additional data related to devices etc.

For example 
A customer has an asset (A factory building) where many temperature sensors are installed.
Along with time series for each sensor, we need to store additional data about devices and asset, 
for example

- Device Id (generated when device was provisioned), device description, manufacturer, installation date etc
- Device location coordinates
- Factory building related meta data

Do ppl use another relational db togather with timeseries db to store this kind of data ?
What is general best practice, how others do it ?

**Comment by mmano** [comment]
Hello,
Tags and attributes should be used as best as possible.
But ... implementation is in my point of view (v1.3) very restrictive. We cannot use them with _select_ clause (enabled only with _show timeseries_ clauses, and with tags, not attributes). I understand it's relevant to indexing and hash map topics .. but it's far from handy. A lot of pre-constructed and initialization stuff could easily cope with performances impact, if any. Even with tags, the _where_ clause is too much limited (only _equal_ or _contains_ operator is accepted, logical construct is not possible). Besides, tagging is not supported with templates. And do not forget that tags/attributes are not versionned, so changes management is very difficult.
My experience ... we end to use another DB for additional (meta)data. Issue is acceptable for internal application development, but may be a showstopper for transversal tooling. It leads to serious tech limitations and dev costs inflation.
I've also set up a couple of "system" timeseries, and I use them in a relational way ... I mean, I use the timestamp as an ID but it's a bit dirty.

To summarize, imho it's not possible to use IoTDB as an unique repository for application with reasonable needs. it could be dependent on the roadmap of IoTDB (which I'm not aware of), but I would advise to only use IoTDB for an efficient raw storage. IoTDB tagging is quite limited in comparison with Influx for example.

**Reply by kershov** [reply]
I'm trying to evaluate Apache IoTDB to fit my needs. Сould you please help me to find out if can do the following: I've got some sensors (e.g. CO, CO2, SO2, etc.) that have some thresholds defined. I'd like not only to query measurements for these sensors, but check if thresholds were exceeded. Is it possible in Apache IoTDB? Or I should use it together with some other DB and somehow do some kind of postprocessing inside my app? Thanks!
